File: mod-common/src/main/java/com/dreamdisplays/net/DeletePacket.java
package com.dreamdisplays.net;

import com.dreamdisplays.PlatformlessInitializer;

import java.util.UUID;
import net.minecraft.core.UUIDUtil;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import org.jspecify.annotations.NullMarked;

// Packet for deleting a display
@NullMarked
public record DeletePacket(UUID id) implements CustomPacketPayload {
    public static final Type<DeletePacket> PACKET_ID =
            new Type<>(Identifier.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "delete"));

    public static final StreamCodec<FriendlyByteBuf, DeletePacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> UUIDUtil.STREAM_CODEC.encode(buf, packet.id()),
                    (buf) -> {
                        UUID id = UUIDUtil.STREAM_CODEC.decode(buf);
                        return new DeletePacket(id);
                    });

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/common/DeletePacket.kt
package com.dreamdisplays.net.common

import com.dreamdisplays.net.common.helpers.createType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type
import java.util.*

data class DeletePacket(val uuid: UUID) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        @JvmField
        val PACKET_ID: Type<DeletePacket> = createType("delete")

        @JvmField
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, DeletePacket> = of(
            { buf, packet -> buf.writeUUID(packet.uuid) },
            { buf -> DeletePacket(buf.readUUID()) }
        )
    }
}
